design/control-data-plane-separation/design.md

```diff
@@ -0,0 +1,770 @@
+
+[gw-tls-config]: https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1beta1.GatewayTLSConfig
+
+[sds]: https://www.envoyproxy.io/docs/envoy/latest/configuration/security/secret
+
+[host-path]: https://kubernetes.io/docs/concepts/storage/volumes/#hostpath
+
+[config-resp]: https://github.com/nginx/agent/blob/ea3a1b4df5d7ecf95bd3d9297d26e420f5e1dd57/sdk/proto/command.pb.go#L891
+
+## Agent Configuration
+
+We can configure the agent through a YAML file, command-line flags, or environment variables on start-up. The agent
+interprets configuration in the following priorities (from highest to lowest):
+
+1. Command-line flags
+2. Environment variables
+3. Config files
+
+We can either configure the agent with command-line flags or mount the config file as a volume to the agent’s container.
+Some configuration options we care about are:
+
+- `server-host`: the DNS name of the control plane Kubernetes Service
+- `server-grpcport`: the gRPC port that the control plane is listening on
+- `tls-ca`: the path to the CA certificate file
+- `tls-cert`: the path to the agent’s certificate file
+- `tls-key`: the path to the agent’s private key file
+- `tls-enable`: enable TLS for secure communication with the control plane
+- `features`: list of enabled features
+- `config-dirs`:the paths that you want to grant agent read/write access to
+
+For a list of configuration options, see [NGINX Agent CLI Flags & Usage][cli].
+
+The control plane can send runtime configuration using the [`AgentConfig`][agent-config] messages. I’m not sure if we
+have a use case for runtime configuration at the moment.
+
+[agent-config]: https://github.com/nginx/agent/blob/ea3a1b4df5d7ecf95bd3d9297d26e420f5e1dd57/sdk/proto/agent.pb.go#L320
+
+[cli]: https://docs.nginx.com/nginx-management-suite/nginx-agent/install-nginx-agent/#nginx-agent-cli-flags-usage
+
+## Data Plane Scaling
+
+Since the data plane is deployed in its own Pod, a user can horizontally scale the data plane independently of the
+control plane. When a new agent Pod spins up, the agent will register itself with the control plane. The control plane
+will have to keep track of all the agents.
+
+We will need to figure out how many agent instances can be supported by NKG’s control plane. As an example data point,
+NGINX Controller can monitor and manage up to 100 N+ instances or up to 30 N+ instances with App Protect enabled. Since
+our use case for the agent is much simpler than NGINX Controller, we may be able to support more instances of the agent.
+
+## Metrics
+
+The agent exposes a REST `/metrics` endpoint that returns metrics in Prometheus format. The agent also sends metrics to
+the control plane. One feature we will need to add to the agent is the ability to enrich the metrics with additional
+information. For example, upstream metrics need to be tied to the Kubernetes Pod name and namespace for them to be
+useful.
+
+## Performance
```

Comment:
Even though we don't have e2e tests yet, but perhaps we can still define scalability goals (number of data plane pods, config. resources, frequency of changes, etc.), so that we can test the implementation manually against them? To confirm that the agent will indeed work for us

Reply:
Sure. I can add those scalability goals to this section of the doc.